diff -rupN obj_orig//cuddObj.cc obj//cuddObj.cc
--- obj_orig//cuddObj.cc	2012-02-02 14:53:32.210106939 +0100
+++ obj//cuddObj.cc	2012-02-02 14:55:53.471668626 +0100
@@ -97,7 +97,7 @@ DD::DD(const DD &from) {
 DD::~DD() {}
 
 
-inline DdManager *
+DdManager *
 DD::checkSameManager(
   const DD &other) const
 {
@@ -110,7 +110,7 @@ DD::checkSameManager(
 } // DD::checkSameManager
 
 
-inline void
+void
 DD::checkReturnValue(
   const DdNode *result) const
 {
@@ -142,7 +142,7 @@ DD::checkReturnValue(
 } // DD::checkReturnValue
 
 
-inline void
+void
 DD::checkReturnValue(
   const int result,
   const int expected) const
@@ -183,14 +183,6 @@ DD::manager() const
 } // DD::manager
 
 
-inline DdNode *
-DD::getNode() const
-{
-    return node;
-
-} // DD::getNode
-
-
 int
 DD::nodeCount() const
 {
@@ -1100,7 +1092,7 @@ Cudd::getHandler() const
 } // Cudd::getHandler
 
 
-inline void
+void
 Cudd::checkReturnValue(
   const DdNode *result) const
 {
@@ -1115,7 +1107,7 @@ Cudd::checkReturnValue(
 } // Cudd::checkReturnValue
 
 
-inline void
+void
 Cudd::checkReturnValue(
   const int result) const
 {
diff -rupN obj_orig//cuddObj.hh obj//cuddObj.hh
--- obj_orig//cuddObj.hh	2012-02-02 14:53:32.210106939 +0100
+++ obj//cuddObj.hh	2012-02-02 14:55:32.638488660 +0100
@@ -103,17 +103,16 @@ class DD {
     friend class ZDD;
     Cudd *ddMgr;
     DdNode *node;
-    inline DdManager * checkSameManager(const DD &other) const;
-    inline void checkReturnValue(const DdNode *result) const;
-    inline void checkReturnValue(const int result, const int expected = 1)
-	const;
+    DdManager * checkSameManager(const DD &other) const;
+    void checkReturnValue(const DdNode *result) const;
+    void checkReturnValue(const int result, const int expected = 1) const;
 public:
     DD(Cudd *ddManager, DdNode *ddNode);
     DD();
     DD(const DD &from);
     virtual ~DD();
     Cudd *manager() const;
-    inline DdNode * getNode() const;
+    inline DdNode * getNode() const { return node; }
     int nodeCount() const;
     unsigned int NodeReadIndex() const;
 
@@ -480,8 +479,8 @@ public:
     inline void makeVerbose() {p->verbose = 1;}
     inline void makeTerse() {p->verbose = 0;}
     inline int isVerbose() const {return p->verbose;}
-    inline void checkReturnValue(const DdNode *result) const;
-    inline void checkReturnValue(const int result) const;
+    void checkReturnValue(const DdNode *result) const;
+    void checkReturnValue(const int result) const;
     Cudd& operator=(const Cudd& right);
     void info() const;
     BDD bddVar();
