diff -crB llvmCore-2.8-rc2.src/tools/clang/include/clang/Parse/Parser.h llvm-src/tools/clang/include/clang/Parse/Parser.h
*** llvmCore-2.8-rc2.src/tools/clang/include/clang/Parse/Parser.h	2010-09-03 03:29:35.000000000 +0200
--- llvm-src/tools/clang/include/clang/Parse/Parser.h	2010-09-28 16:03:41.068158812 +0200
***************
*** 24,29 ****
--- 24,33 ----
  #include <stack>
  #include <list>
  
+ //--- INSIEME COMPILER ---//
+ class ParserProxy;
+ //-----------------------//
+ 
  namespace clang {
    class AttributeList;
    struct CXX0XAttributeList;
***************
*** 76,81 ****
--- 80,88 ----
    friend class PragmaUnusedHandler;
    friend class ColonProtectionRAIIObject;
    friend class ParenBraceBracketBalancer;
+   //--- INSIEME COMPILER ---//
+ 	friend class ::ParserProxy;
+ //-----------------------//
    PrettyStackTraceParserEntry CrashInfo;
  
    Preprocessor &PP;
diff -crB llvmCore-2.8-rc2.src/tools/clang/include/clang/Sema/Sema.h llvm-src/tools/clang/include/clang/Sema/Sema.h
*** llvmCore-2.8-rc2.src/tools/clang/include/clang/Sema/Sema.h	2010-09-03 04:10:08.000000000 +0200
--- llvm-src/tools/clang/include/clang/Sema/Sema.h	2010-09-28 16:08:09.348168358 +0200
***************
*** 652,658 ****
                                 CXXScopeSpec *SS,
                                 ParsedType &SuggestedType);
  
!   Decl *ActOnDeclarator(Scope *S, Declarator &D);
  
    Decl *HandleDeclarator(Scope *S, Declarator &D,
                           MultiTemplateParamsArg TemplateParameterLists,
--- 652,658 ----
                                 CXXScopeSpec *SS,
                                 ParsedType &SuggestedType);
  
!   virtual Decl *ActOnDeclarator(Scope *S, Declarator &D);
  
    Decl *HandleDeclarator(Scope *S, Declarator &D,
                           MultiTemplateParamsArg TemplateParameterLists,
***************
*** 722,732 ****
                                           unsigned NumDecls);
    void ActOnFinishKNRParamDeclarations(Scope *S, Declarator &D,
                                         SourceLocation LocAfterDecls);
!   Decl *ActOnStartOfFunctionDef(Scope *S, Declarator &D);
!   Decl *ActOnStartOfFunctionDef(Scope *S, Decl *D);
    void ActOnStartOfObjCMethodDef(Scope *S, Decl *D);
  
!   Decl *ActOnFinishFunctionBody(Decl *Decl, Stmt *Body);
    Decl *ActOnFinishFunctionBody(Decl *Decl, Stmt *Body, bool IsInstantiation);
  
    /// \brief Diagnose any unused parameters in the given sequence of
--- 722,732 ----
                                           unsigned NumDecls);
    void ActOnFinishKNRParamDeclarations(Scope *S, Declarator &D,
                                         SourceLocation LocAfterDecls);
!   virtual Decl *ActOnStartOfFunctionDef(Scope *S, Declarator &D);
!   virtual Decl *ActOnStartOfFunctionDef(Scope *S, Decl *D);
    void ActOnStartOfObjCMethodDef(Scope *S, Decl *D);
  
!   virtual Decl *ActOnFinishFunctionBody(Decl *Decl, Stmt *Body);
    Decl *ActOnFinishFunctionBody(Decl *Decl, Stmt *Body, bool IsInstantiation);
  
    /// \brief Diagnose any unused parameters in the given sequence of
***************
*** 812,818 ****
                    tok::ObjCKeywordKind visibility);
  
    // This is used for both record definitions and ObjC interface declarations.
!   void ActOnFields(Scope* S, SourceLocation RecLoc, Decl *TagDecl,
                     Decl **Fields, unsigned NumFields,
                     SourceLocation LBrac, SourceLocation RBrac,
                     AttributeList *AttrList);
--- 812,818 ----
                    tok::ObjCKeywordKind visibility);
  
    // This is used for both record definitions and ObjC interface declarations.
!   virtual void ActOnFields(Scope* S, SourceLocation RecLoc, Decl *TagDecl,
                     Decl **Fields, unsigned NumFields,
                     SourceLocation LBrac, SourceLocation RBrac,
                     AttributeList *AttrList);
***************
*** 830,836 ****
  
    /// ActOnTagFinishDefinition - Invoked once we have finished parsing
    /// the definition of a tag (enumeration, class, struct, or union).
!   void ActOnTagFinishDefinition(Scope *S, Decl *TagDecl,
                                  SourceLocation RBraceLoc);
  
    /// ActOnTagDefinitionError - Invoked when there was an unrecoverable
--- 830,836 ----
  
    /// ActOnTagFinishDefinition - Invoked once we have finished parsing
    /// the definition of a tag (enumeration, class, struct, or union).
!   virtual void ActOnTagFinishDefinition(Scope *S, Decl *TagDecl,
                                  SourceLocation RBraceLoc);
  
    /// ActOnTagDefinitionError - Invoked when there was an unrecoverable
***************
*** 1513,1519 ****
    StmtResult ActOnExprStmt(FullExprArg Expr);
  
    StmtResult ActOnNullStmt(SourceLocation SemiLoc);
!   StmtResult ActOnCompoundStmt(SourceLocation L, SourceLocation R,
                                         MultiStmtArg Elts,
                                         bool isStmtExpr);
    StmtResult ActOnDeclStmt(DeclGroupPtrTy Decl,
--- 1513,1519 ----
    StmtResult ActOnExprStmt(FullExprArg Expr);
  
    StmtResult ActOnNullStmt(SourceLocation SemiLoc);
!   virtual StmtResult ActOnCompoundStmt(SourceLocation L, SourceLocation R,
                                         MultiStmtArg Elts,
                                         bool isStmtExpr);
    StmtResult ActOnDeclStmt(DeclGroupPtrTy Decl,
***************
*** 1532,1538 ****
                                      IdentifierInfo *II,
                                      SourceLocation ColonLoc,
                                      Stmt *SubStmt);
!   StmtResult ActOnIfStmt(SourceLocation IfLoc,
                                   FullExprArg CondVal, Decl *CondVar,
                                   Stmt *ThenVal,
                                   SourceLocation ElseLoc, Stmt *ElseVal);
--- 1532,1538 ----
                                      IdentifierInfo *II,
                                      SourceLocation ColonLoc,
                                      Stmt *SubStmt);
!   virtual StmtResult ActOnIfStmt(SourceLocation IfLoc,
                                   FullExprArg CondVal, Decl *CondVar,
                                   Stmt *ThenVal,
                                   SourceLocation ElseLoc, Stmt *ElseVal);
***************
*** 1549,1555 ****
                                   SourceLocation CondLParen, Expr *Cond,
                                   SourceLocation CondRParen);
  
!   StmtResult ActOnForStmt(SourceLocation ForLoc,
                            SourceLocation LParenLoc,
                            Stmt *First, FullExprArg Second,
                            Decl *SecondVar,
--- 1549,1555 ----
                                   SourceLocation CondLParen, Expr *Cond,
                                   SourceLocation CondRParen);
  
!   virtual StmtResult ActOnForStmt(SourceLocation ForLoc,
                            SourceLocation LParenLoc,
                            Stmt *First, FullExprArg Second,
                            Decl *SecondVar,
